File: framework/languages/java/samples/java/DeliveryDispatch/Server/Tracking/src/main/java/systems/zlink/samples/deliverydispatch/server/tracking/handlers/SubscribeCustomerToDeliveryHandler.java
package systems.zlink.samples.deliverydispatch.server.tracking.handlers;

import com.fasterxml.jackson.databind.ObjectMapper;
import systems.zlink.contracts.core.RoutingId;
import systems.zlink.framework.ZLinkAwait;
import systems.zlink.framework.actors.ZLinkActorGateway;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.framework.spots.ZLinkSpotManager;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleNames;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleTimings;
import systems.zlink.samples.deliverydispatch.server.tracking.spots.DeliveryTrackingSpot;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

@ZLinkHandlerGroup("tracking")
public final class SubscribeCustomerToDeliveryHandler
    implements ZLinkRequestHandler<
        Messages.SubscribeCustomerToDelivery,
        Messages.CustomerDeliverySubscribed> {
    private final ZLinkActorManager actors;
    private final ZLinkActorGateway actorGateway;
    private final ZLinkSpotManager spots;
    private final ObjectMapper json;

    public SubscribeCustomerToDeliveryHandler(
        ZLinkActorManager actors,
        ZLinkActorGateway actorGateway,
        ZLinkSpotManager spots,
        ObjectMapper json) {
        this.actors = actors;
        this.actorGateway = actorGateway;
        this.spots = spots;
        this.json = json;
    }

    @Override
    public Messages.CustomerDeliverySubscribed handle(
        Messages.SubscribeCustomerToDelivery request,
        ZLinkRequestContext context) {
        ZLinkAwait.await(spots.getOrCreate(
            DeliveryTrackingSpot.class,
            RoutingId.from(request.deliveryId()),
            new Messages.DeliverySpotCreate(request.deliveryId())));
        var actor = ZLinkAwait.await(
            actors.getOrCreate(request.customerId(), SampleNames.CustomerActorType));
        actorGateway
            .joinSpot(
                actor,
                RoutingId.from(request.deliveryId()),
                new Messages.DeliverySpotJoin(request.deliveryId(), request.customerId()))
            .timeout(SampleTimings.RequestTimeout)
            .await(Messages.DeliverySpotJoined.class);
        return new Messages.CustomerDeliverySubscribed(request.customerId(), request.deliveryId());
    }

}

File: framework/languages/java/samples/kotlin/DeliveryDispatch/Server/Tracking/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/tracking/handlers/SubscribeCustomerToDeliveryHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.tracking.handlers

import com.fasterxml.jackson.databind.ObjectMapper
import kotlinx.coroutines.future.await
import systems.zlink.contracts.core.RoutingId
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.channels.ZLinkRequestContext
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.spots.ZLinkSpotManager
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleNames
import systems.zlink.samples.kotlin.deliverydispatch.server.tracking.spots.deliverytrackingspot.DeliveryTrackingSpot
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.SubscribeCustomerToDeliveryRes
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliverySpotCreateReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.SubscribeCustomerToDeliveryReq

@ZLinkHandlerGroup("tracking")
class SubscribeCustomerToDeliveryHandler(
    private val actors: ZLinkActorManager,
    private val spots: ZLinkSpotManager,
    private val json: ObjectMapper,
) : ZLinkSuspendingRequestHandler<SubscribeCustomerToDeliveryReq, SubscribeCustomerToDeliveryRes> {
    override suspend fun handle(
        request: SubscribeCustomerToDeliveryReq,
        context: ZLinkRequestContext,
    ) = run {
        spots.getOrCreate(
            DeliveryTrackingSpot::class.java,
            RoutingId.from(request.deliveryId),
            DeliverySpotCreateReq(request.deliveryId),
        ).await()
        val actor = actors.getOrCreate(request.customerId, SampleNames.CustomerActorType).await()
        require(actor.actorId().isNotBlank())
        SubscribeCustomerToDeliveryRes(request.customerId, request.deliveryId)
    }
}
